Xamarin.Essentials/WebAuthenticator/WebAuthenticator.android.cs

```diff
@@ -0,0 +1,130 @@
+﻿using System;
+using System.Threading.Tasks;
+using Android.Content;
+using Android.Support.CustomTabs;
+
+namespace Xamarin.Essentials
+{
+    public partial class WebAuthenticator
+    {
+        static TaskCompletionSource<WebAuthenticatorResult> tcsResponse = null;
+
+        static Uri uri = null;
+
+        static CustomTabsActivityManager CustomTabsActivityManager { get; set; }
+
+        static Uri RedirectUri { get; set; }
+
+        internal static Task<WebAuthenticatorResult> ResponseTask
+            => tcsResponse?.Task;
+
+        internal static bool OnResume(Intent intent)
+        {
+            // If we aren't waiting on a task, don't handle the url
+            if (tcsResponse?.Task?.IsCompleted ?? true)
+                return false;
+
+            if (intent == null)
+            {
+                tcsResponse.TrySetCanceled();
+                return false;
+            }
+
+            try
+            {
+                var intentUri = new Uri(intent.Data.ToString());
+
+                // Only handle schemes we expect
+                if (!WebUtils.CanHandleCallback(RedirectUri, intentUri))
+                {
+                    tcsResponse.TrySetException(new InvalidOperationException($"Invalid Redirect URI, detected `{intentUri}` but expected a URI in the format of `{RedirectUri}`"));
+                    return false;
+                }
+
+                tcsResponse?.TrySetResult(new WebAuthenticatorResult(intentUri));
+                return true;
+            }
+            catch (Exception ex)
+            {
+                tcsResponse.TrySetException(ex);
+                return false;
+            }
+        }
+
+        static Task<WebAuthenticatorResult> PlatformAuthenticateAsync(Uri url, Uri callbackUrl)
+        {
+            var packageName = Platform.AppContext.PackageName;
+
+            // Create an intent to see if the app developer wired up the callback activity correctly
+            var intent = new Intent(Intent.ActionView);
+            intent.AddCategory(Intent.CategoryBrowsable);
+            intent.AddCategory(Intent.CategoryDefault);
+            intent.SetPackage(packageName);
+            intent.SetData(global::Android.Net.Uri.Parse(callbackUrl.OriginalString));
+
+            // Try to find the activity for the callback intent
+            var c = intent.ResolveActivity(Platform.AppContext.PackageManager);
+
+            if (c == null || c.PackageName != packageName)
+                throw new InvalidOperationException($"You must subclass the `{nameof(WebAuthenticatorCallbackActivity)}` and create an IntentFilter for it which matches your `{nameof(callbackUrl)}`.");
```

Comment:
I love this. Smart and helpful. Before even starting the auth, make sure the activity exists.

Reply:
This caused me some fits because the sample code using a custom scheme name, while I am authenticating against an oauth2 endpoint that only supports http or https.  After copying that code to my own android specific handler and running a ton of experiments, I was able to finally get it working by specifying DataScheme and DataHost both.  I added DataPath as well, just to make it more specific to the use case finding it required / before and after to match.

  [Activity(NoHistory = true, LaunchMode = LaunchMode.SingleTop)]
  [IntentFilter(
    new[] { Android.Content.Intent.ActionView },
    Categories = new[] { Android.Content.Intent.CategoryDefault, Android.Content.Intent.CategoryBrowsable },
    DataScheme = "https",
    DataHost = "my-company.com",
    DataPath = "/auth/")]
  public sealed class WebAuthenticatorActivity : Xamarin.Essentials.WebAuthenticatorCallbackActivity
  {